src/corefx/System.Globalization.Native/locale.cpp

```diff
@@ -0,0 +1,373 @@
+	}
+
+	return loc;
+}
+
+/*
+Function:
+GopyCharToUChar
+
+Copies the given char to UChar with error checking
+*/
+UErrorCode CopyCharToUChar(const char *str, UChar* value, int32_t valueLength)
+{
+	int len = strlen(str);
+	if (len >= valueLength)
+	{
+		return U_BUFFER_OVERFLOW_ERROR;
+	}
+	u_charsToUChars(str, value, len + 1);
+	return U_ZERO_ERROR;
+}
+
+/*
+PAL Function:
+getCanonicalLocaleName
+
+Obtains a canonical locale given the locale name
+*/
+extern "C" int32_t GetLocaleName(const UChar* localeName, UChar* value, int32_t valueLength)
+{
+	Locale locale = GetLocale(localeName, true);
+
+	if (locale.isBogus())
+	{
+		return U_ILLEGAL_ARGUMENT_ERROR;
+	}
+
+	if (strlen(locale.getISO3Language()) == 0)
+	{
+		// unknown language; language is required (script and country optional)
+		return U_ILLEGAL_ARGUMENT_ERROR;
+	}
+
+	UErrorCode status = CopyCharToUChar(locale.getName(), value, valueLength);
+	if (U_SUCCESS(status))
+	{
+		// replace underscores with hyphens to interop with existing .NET code
+		for (UChar* ch = value; *ch != (UChar)'\0'; ch++)
+		{
+			if (*ch == (UChar)'_')
+			{
+				*ch = (UChar)'-';
+			}
+		}
+	}
+
+	return status;
+}
+
+/*
+Function:
+GetMeasurementSystem
+
+Obtains the measurement system for the local, determining if US or metric
+*/
+UErrorCode GetMeasurementSystem(const char *localeId, int32_t *value)
+{
+	UErrorCode status = U_ZERO_ERROR;
+
+	UMeasurementSystem measurementSystem = ulocdata_getMeasurementSystem(localeId, &status);
+	if (U_SUCCESS(status))
+	{
+		*value = (measurementSystem == UMeasurementSystem::UMS_US) ? 1 : 0;
+	}
+
+	return status;
+}
+
+/*
+PAL Function:
+GetLocaleInfoInt
+
+Returns integer locale information
+*/
+extern "C" int32_t GetLocaleInfoInt(const UChar* localeName, LocaleNumberData localeNumberData, int32_t* value)
+{
+	Locale locale = GetLocale(localeName);
+	if (locale.isBogus())
+	{
+		return U_ILLEGAL_ARGUMENT_ERROR;
+	}
+
+	UErrorCode status = U_ZERO_ERROR;
+	switch (localeNumberData)
+	{
+		case LanguageId:
+			*value = locale.getLCID();
+			break;
+		case MeasurementSystem:
+			status = GetMeasurementSystem(locale.getName(), value);
+			break;
+		case FractionalDigitsCount:
+			*value = 2; //TODO: implement
+			break;
+		case NegativeNumberFormat:
+			*value = 1; //TODO: implement
+			break;
+		case MonetaryFractionalDigitsCount:
+			*value = 2; //TODO: implement
+			break;
+		case PositiveMonetaryNumberFormat:
+			*value = 0; //TODO: implement
+			break;
+		case NegativeMonetaryNumberFormat:
+			*value = 0; //TODO: implement
+			break;
+		case CalendarType:
+			*value = 1; //TODO: implement
+			break;
+		case FirstWeekOfYear:
+			*value = 0; //TODO: implement
+			break;
+		case ReadingLayout:
+			*value = 0; //TODO: implement
+			break;
+		case NegativePercentFormat:
+			*value = 0; //TODO: implement
+			break;
+		case PositivePercentFormat:
+			*value = 0; //TODO: implement
+			break;
+		default:
+			status = U_UNSUPPORTED_ERROR;
```

Comment:
I see now in some cases you have asserts on the managed side to catch this, but I think it would be nice to have an assert on the native side instead or as well.